APIdefinition.md

```diff
@@ -0,0 +1,75 @@
+# ISIS Project public API
+
+The public API is considered to be all the interfaces that the ISIS
+project provides to the community from which the ISIS user base can
+build more complex tools. This API therefore includes code-level
+structures accessible via software calls to libraries, but also
+more user-facing structures like command-line programs and graphical
+user interface (GUI) programs. Any "interface" via which a user can
+reasonably interact with components of the ISIS project should be
+considered part of the "ISIS Project public API" for the purposes
+of [Semantic Versioning](https://semver.org).
+
+Examples are provided below, but are not meant to be exhaustive or
+complete. The above philosophy should guide our thinking in what
+the public API is.
+
+
+## General Guidance
+
+**Code-level structures**: Function, class, and object names,
+argument signatures, and return signatures for functions, classes,
+and objects.
+
+- C++ code exceptions would be those structures above marked with
+  a "private" access specifier via the C++ language.
+- Python code exceptions would be those structures above which begin 
+  with an underbar ("_", as per Python language tradition).
+
+**User-facing command-line and GUI programs**: program name, argument
+list, argument defaults, configuration or preference names and their
+defaults, and for those programs with structured output files, the
+structure and contents of those output files are analogous to the
```

Comment:
I would say structure, but not contents. We can say that the existence of a keyword is part of the public API, but the contents of the keyword is not.

Reply:
I disagree, perhaps because we have sliced the meaning of "contents" differently, I think it is important to keep both words in that sentence to make people think about possibilities, and I think they are mitigated by the presence of "are analogous" later on and the specific examples farther down the document.

I would agree with you that the value of a keyword is not at issue, but as in my PVL example, the simple enclosure of a number in quotes changes the "type" of that value from an int to a string.  You might conceive of that as "structure," but I consider it "contents" because it is the data type of the "contents" of the assigned variable.  If a function returned a dict, and the value of the key "foo" changed from int to string, that would be a problem, and that value is the "contents" of the dict, but maybe it is structure?

So we could go deep in this first para and strictly and exhaustively define terms, or we could leave it to the file-type examples later in the document to provide specific guidance about particular file types.

Reply:
I agree with @jessemapel here. The word 'content' is way too ambiguous. One could argue that the information content of 2 and "2" is identical. We all just read two. If this is about types, then we need to say types. If this is about semantic content, then we are talking about values which do not, in my opinion, belong in the API.

Reply:
A human might argue that they are identical, but a standards-compliant PVL parser would interpret those differently.  I'm happy to strike "content" here as being potentially confusing, and allow us to be potentially more specific with later examples.